codex-rs/protocol/src/models.rs

```diff
@@ -309,6 +314,63 @@ impl std::ops::Deref for FunctionCallOutputPayload {
     }
 }
 
+// Lightweight, lossless mapping from model `ResponseItem` values to one or more
+// `EventMsg` payloads that UIs can render directly. Variants that require
+// side-effects (e.g., FunctionCall, LocalShellCall, CustomToolCall) do not
+// emit any events here and are handled by higher layers.
+impl From<&ResponseItem> for Vec<EventMsg> {
+    fn from(item: &ResponseItem) -> Self {
+        let mut events: Vec<EventMsg> = Vec::new();
+
+        if let ResponseItem::Message { content, .. } = item {
```

Comment:
Prefer `match` with guards so this function is forced to be updated if a new variant of `ResponseItem` is added.